### PR TITLE
Add the `CACert` parameter to the `ExternalCA` object

### PR DESCRIPTION
**Pull request by cyli**
Add the `CACert` parameter the `ExternalCA` object  in order to match
swarmkit's API type.  Make sure this parameter gets propagated to
swarmkit, and also add an extra option to the CLI when providing
external CAs to parse the CA cert from a file.

Signed-off-by: Ying Li <ying.li@docker.com>

This is a follow-up to https://github.com/moby/moby/issues/32576 (see https://github.com/moby/moby/pull/32563#issuecomment-293700378 for reference)

image

cc @diogomonica @aaronlehmann

### CODE REVIEW

File: integration-cli/docker_api_swarm_test.go [hunk 1]
```diff
@@ -157,11 +156,18 @@ func (s *DockerSwarmSuite) TestUpdateSwarmAddExternalCA(c *check.C) {
...
+				CACert:   []byte("cacert"),
+			},
 		}
 	})
 	info, err := d1.SwarmInfo()
 	c.Assert(err, checker.IsNil)
-	c.Assert(info.Cluster.Spec.CAConfig.ExternalCAs, checker.HasLen, 1)
+	c.Assert(info.Cluster.Spec.CAConfig.ExternalCAs, checker.HasLen, 2)
+	c.Assert(info.Cluster.Spec.CAConfig.ExternalCAs[0].CACert, checker.IsNil)
+	c.Assert(reflect.DeepEqual(info.Cluster.Spec.CAConfig.ExternalCAs[1].CACert, []byte("cacert")), checker.True)
```
Comment (aaronlehmann) [inline integration-cli/docker_api_swarm_test.go]: Does `check.Equals` not support comparing byte slices? If not, I think `check.DeepEquals` would work.

**Comment by aaronlehmann** [comment]
Code changes look good, but please update `api/swagger.yaml`.

File: integration-cli/docker_cli_swarm_test.go [hunk 2]
```diff
@@ -68,14 +77,22 @@ func (s *DockerSwarmSuite) TestSwarmInit(c *check.C) {
...
 	spec := getSpec()
 	c.Assert(spec.CAConfig.NodeCertExpiry, checker.Equals, 30*time.Hour)
 	c.Assert(spec.Dispatcher.HeartbeatPeriod, checker.Equals, 11*time.Second)
-	c.Assert(spec.CAConfig.ExternalCAs, checker.HasLen, 1)
+	c.Assert(spec.CAConfig.ExternalCAs, checker.HasLen, 2)
+	c.Assert(spec.CAConfig.ExternalCAs[0].CACert, checker.IsNil)
+	c.Assert(reflect.DeepEqual(spec.CAConfig.ExternalCAs[1].CACert, []byte("fakecert")), checker.True)
 
 	c.Assert(d.Leave(true), checker.IsNil)
 	time.Sleep(500 * time.Millisecond) // https://github.com/docker/swarmkit/issues/1421
```
Comment (aaronlehmann) [inline integration-cli/docker_cli_swarm_test.go]: It should be safe to remove this now, since the underlying swarmkit issue (actually a etcd/raft issue) has been fixed. We can do it in a different PR though if you prefer that. Let me know if I should open one.

File: integration-cli/docker_api_swarm_test.go (hunk 1 above)
Comment (cyli) [inline integration-cli/docker_api_swarm_test.go]: `check.Equals` does not.   I didn't see `check.DeepEquals` though - will use that instead.

File: integration-cli/docker_cli_swarm_test.go (hunk 2 above)
Comment (cyli) [inline integration-cli/docker_cli_swarm_test.go]: I'm ok with doing this here, since I'm updating anyway :)

**Comment by cyli** [comment]
@aaronlehmann Added the parameter to `api/swagger.yaml` - I wasn't entirely sure what type `[]byte` was, so I just copied the value from `SecretSpec.data` - `type: "array"` with `items: type: "string"`.

I notice that `RuntimeData` should also be `[]byte`, but its type is just `type: "array"` (without the `item` type.  Is there a preference as to which it should be?

**Comment by aaronlehmann** [comment]
Note that `[]byte` is automatically base64 encoded/decoded by Go's JSON marshaller and unmarshaller. I think these certs are PEM armored and therefore another layer of encoding would be unnecessary. Should they be strings instead in the REST API to avoid base64 encoding?

**Comment by cyli** [comment]
@aaronlehmann We don't validate that the contents of the file we read are PEM certificates - we just pass the bytes on directly.  I can convert the bytes to string, but if it's invalid file (that's DER format for example), would that cause issues?  I could also just add another check to make sure it's a valid PEM object when parsing.

File: integration-cli/docker_cli_swarm_test.go [hunk 3]
```diff
@@ -52,12 +52,20 @@ func (s *DockerSwarmSuite) TestSwarmUpdate(c *check.C) {
 	spec = getSpec()
 	c.Assert(spec.CAConfig.NodeCertExpiry, checker.Equals, 30*time.Hour)
 
+	tempFile, err := ioutil.TempFile("", "testfile")
+	c.Assert(err, checker.IsNil)
+	defer os.Remove(tempFile.Name())
+	tempFile.Write([]byte("fakecert"))
```
Comment (dnephin) [inline integration-cli/docker_cli_swarm_test.go]:
```go
import "github.com/pkg/docker/docker/testutil/tempfile"
...


tempFile, err := tempfile.NewTempFile(c, "testfile", "fakecert")
defer tempFile.Remove()

```

File: integration-cli/docker_cli_swarm_test.go [hunk 4]
```diff
@@ -68,17 +76,24 @@ func (s *DockerSwarmSuite) TestSwarmInit(c *check.C) {
 		return sw.Spec
 	}
 
+	tempFile, err := ioutil.TempFile("", "testfile")
+	c.Assert(err, checker.IsNil)
+	defer os.Remove(tempFile.Name())
+	tempFile.Write([]byte("fakecert"))
```
Comment (dnephin) [inline integration-cli/docker_cli_swarm_test.go]: `tempfile.NewTempFile()`

File: integration-cli/docker_cli_swarm_test.go [hunk 5]
```diff
@@ -52,12 +52,20 @@ func (s *DockerSwarmSuite) TestSwarmUpdate(c *check.C) {
...
+
 	// passing an external CA (this is without starting a root rotation) does not fail
-	out, err = d.Cmd("swarm", "update", "--external-ca", "protocol=cfssl,url=https://something.org")
+	out, err = d.Cmd("swarm", "update", "--external-ca", "protocol=cfssl,url=https://something.org",
+		"--external-ca", fmt.Sprintf("protocol=cfssl,url=https://somethingelse.org,cacert=%s", tempFile.Name()))
```
Comment (dnephin) [inline integration-cli/docker_cli_swarm_test.go]: Using this will give much better error messages when there is a failure:

```go
result := cli.Docker(cli.Args(...), cli.Daemon(d.Daemon))
result.Assert(c, icmd.Success)
```

**Review by dnephin — COMMENTED** [review]
Just a couple small comments about the tests, otherwise looks good

File: integration-cli/docker_cli_swarm_test.go (hunk 3 above)
Comment (cyli) [inline integration-cli/docker_cli_swarm_test.go]: Done, thanks!

File: integration-cli/docker_cli_swarm_test.go (hunk 5 above)
Comment (cyli) [inline integration-cli/docker_cli_swarm_test.go]: Fixed

File: integration-cli/docker_cli_swarm_test.go (hunk 4 above)
Comment (cyli) [inline integration-cli/docker_cli_swarm_test.go]: Fixed, thanks :)

**Comment by cyli** [comment]
@aaronlehmann Changed to `string` instead of `[]byte`, and in the CLI validate that it's a PEM block.

File: cli/command/swarm/opts.go [hunk 6]
```diff
@@ -3,9 +3,12 @@ package swarm
 import (
 	"encoding/csv"
 	"fmt"
+	"io/ioutil"
 	"strings"
 	"time"
```
Comment (aaronlehmann) [inline cli/command/swarm/opts.go]: There's an extra blank line here

**Comment by aaronlehmann** [comment]
LGTM

File: cli/command/swarm/opts.go (hunk 6 above)
Comment (cyli) [inline cli/command/swarm/opts.go]: Left one blank line between the stdlib imports and the `github.com/*` imports, but removed the extraneous ones between the stdlib imports.

**Comment by diogomonica** [comment]
LGTM

**Review by dnephin — APPROVED** [review]
LGTM
